### PR TITLE
fix: Using batchsize of batch_sampler when using BatchSampler

### LINKED ISSUE #21122: batch sampler, ddp, and sampler wrapper
### Bug description

I might be wrong, but when using a BatchSampler with DDP, Lightning never wraps it with a DistributedSamplerWrapper. This is due to the following condition in its code:
```
if isinstance(dataloader.sampler, (RandomSampler, SequentialSampler)):
    return DistributedSampler(dataloader.dataset, **kwargs)
```
Since PyTorch uses a SequentialSampler by default, this condition is always met
```
DataLoader(ds, sampler=None, batch_sampler=MNSampler(ds.labels)).sampler
<torch.utils.data.sampler.SequentialSampler at 0x7deb78e01eb0>
```

### What version are you seeing the problem on?

v2.4

### Reproduced in studio

_No response_

### How to reproduce the bug

```python

```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by bigd4**
## What does this PR do?

Fixes #21122

When `DataLoader` is defined with `DataLoader(dataset, batch_sampler=BatchSampler(...))`, the `batch_size` and `drop_last` will be default values (1 and False). So `_dataloader_init_kwargs_resolve_sampler` may return wrong parameters.

### Changes
return `batch_sampler.batch_size` and `batch_sampler.drop_last` if `batch_sampler` is `BatchSampler`.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21566.org.readthedocs.build/en/21566/

### CODE REVIEW

File: tests/tests_pytorch/trainer/test_batch_sampler.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    def on_train_start(self, trainer, pl_module):
+        assert isinstance(trainer.train_dataloader.sampler, DistributedSampler)
+        assert trainer.train_dataloader.batch_size == self.expected_batch_size
+        assert trainer.train_dataloader.drop_last == self.expected_drop_last
+
+
+@pytest.mark.parametrize("batch_size", [1, 5])
+@pytest.mark.parametrize("drop_last", [False, True])
+@RunIf(min_cuda_gpus=2, skip_windows=True)
```
Comment (deependujha) [inline tests/tests_pytorch/trainer/test_batch_sampler.py]:
```suggestion
@RunIf(min_cuda_gpus=2, skip_windows=True, standalone=True)
```

make it `standalone`

File: src/lightning/pytorch/utilities/data.py [hunk 2]
```diff
@@ -332,6 +332,17 @@ def _dataloader_init_kwargs_resolve_sampler(
             "batch_size": 1,
             "drop_last": False,
         }
+    if batch_sampler is not None and batch_sampler_cls is BatchSampler:
+        # This is a PyTorch `BatchSampler` but maybe created by user, so batch_size and drop_last should be preserved
+        batch_size = batch_sampler.batch_size
+        drop_last = batch_sampler.drop_last if not is_predicting else False
```
Comment (deependujha) [inline src/lightning/pytorch/utilities/data.py:338]: why `is_predicting` matters?

File: tests/tests_pytorch/trainer/test_batch_sampler.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    def on_train_start(self, trainer, pl_module):
+        assert isinstance(trainer.train_dataloader.sampler, DistributedSampler)
+        assert trainer.train_dataloader.batch_size == self.expected_batch_size
+        assert trainer.train_dataloader.drop_last == self.expected_drop_last
+
+
+@pytest.mark.parametrize("batch_size", [1, 5])
+@pytest.mark.parametrize("drop_last", [False, True])
+@RunIf(min_cuda_gpus=2, skip_windows=True, standalone=True)
```
Comment (deependujha) [inline tests/tests_pytorch/trainer/test_batch_sampler.py]: why is this test in a separate file?

File: src/lightning/pytorch/utilities/data.py (hunk 2 above)
Comment (bigd4) [reply src/lightning/pytorch/utilities/data.py:338]: I am not sure, this is copied from above. Maybe forcing `drop_last=False` in prediction mode ensures no inputs are silently skipped?

File: tests/tests_pytorch/trainer/test_batch_sampler.py (hunk 3 above)
Comment (bigd4) [reply tests/tests_pytorch/trainer/test_batch_sampler.py]: Should I move this to `test_dataloader.py`?

Comment (deependujha) [reply tests/tests_pytorch/trainer/test_batch_sampler.py]: let's do it

File: tests/tests_pytorch/trainer/test_dataloaders.py [hunk 4]
```diff
@@ -814,6 +814,41 @@ def test_dataloader_distributed_sampler(tmp_path):
...
+@pytest.mark.parametrize("drop_last", [False, True])
+@RunIf(min_cuda_gpus=2, skip_windows=True, standalone=True)
+def test_dataloader_distributed_batch_sampler(tmp_path, batch_size, drop_last):
+    """Test BatchSampler and it's arguments for DDP backend."""
+    seed_everything(123)
+    dataset = RandomDataset(32, 64)
+    sampler = RandomSampler(dataset)
+    batch_sampler = BatchSampler(sampler, batch_size=batch_size, drop_last=drop_last)
+    dataloader = DataLoader(dataset, batch_sampler=batch_sampler)
+    print(batch_sampler.drop_last, dataloader.drop_last)
```
Comment (deependujha) [inline tests/tests_pytorch/trainer/test_dataloaders.py]:
```suggestion
```

**Comment by bigd4** [comment]
@deependujha 
The tests fail becase `pytorch-lightning/tests/tests_pytorch/trainer/connectors/test_data_connector.py` line 198-200. The missing value are `['batch_sampler', 'batch_size', 'drop_last', 'sampler', 'shuffle']` instead of `['batch_sampler', 'sampler', 'shuffle']`.
```
    loader = TestDataLoader(ds)
    sampler = SequentialSampler(ds)
    match = escape("missing arguments are ['batch_sampler', 'sampler', 'shuffle']")
    with pytest.raises(MisconfigurationException, match=match):
        _update_dataloader(loader, sampler, mode="fit")
    match = escape("missing arguments are ['batch_sampler', 'batch_size', 'drop_last', 'sampler', 'shuffle']")
    with pytest.raises(MisconfigurationException, match=match):
        _update_dataloader(loader, sampler, mode="predict")
```
I dont know if I should change `match = escape("missing arguments are ['batch_sampler', 'sampler', 'shuffle']")` or any other ideas?
